server-spi-private/src/main/java/org/keycloak/credential/hash/Pbkdf2PasswordHashProvider.java

```diff
@@ -132,4 +136,14 @@ private SecretKeyFactory getSecretKeyFactory() {
             throw new RuntimeException("PBKDF2 algorithm not found", e);
         }
     }
+
+    private String pbkdf2PasswordPadding(String rawPassword) {
+        if (rawPassword.length() < maxPaddingLength) {
+            int nPad = maxPaddingLength - rawPassword.length();
+            String result = rawPassword;
+            for (int i = 0 ; i < nPad; i++) result += "\0";
+            return result;
+        } else
+            return rawPassword;
+    }
```

Comment:
Would have liked to see something testing this padding stuff. What about making it a static method, extracting into a separate class, and adding a unit test?

Reply:
There is unit test `FIPS1402Pbkdf2PasswordPaddingTest` in this PR. It tests various scenarios like backwards compatibility (password without padding is possible to verify with padding) and also check that it works to switch between non-approved and approved mode for password verifications etc.

It is in the `crypto/fips1402` module, and hence only in the FIPS specific module, however I consider this OK due the fact:
- Padding will be most likely used especially by guys in the FIPS mode, if they cannot rely just on password policies (for example due the backwards compatibility and requirement of verification of old passwords)
- These tests fare executed during our regular GH actions build, which runs the unit tests on the `crypto/fips1402` module.

Isn't it sufficient?

Reply:
The tests now as I understand it checks that the password is compatible with padding on/off, and that it passes for the BC provider.

However, it would be nice to see a simple unit test (or something) that just verifies the padding of the string works as expected before the hashing is performed. Could for instance be that it actually pads it to a few characters longer than it should, etc.